File: tools/ts/typesafe.mjs
// Runs the same System One calls through @typesafe-ai/sdk or @kime/sdk and writes the answers and
// the time each call took, so the two clients can be compared answer for answer.
//
//   node tools/ts/typesafe.mjs run <sdk module> <base url> texts.json out.json
//   node tools/ts/typesafe.mjs compare out-a.json out-b.json
//
// <sdk module> is a path to the package entry, such as node_modules/@typesafe-ai/sdk/dist/index.mjs
// or ts/dist/index.js. Each text is asked a noul, a choice and a score question, the first call is
// left out of the timings, and the client overhead is timed on 2000 calls to GET /v1/models.

import { readFileSync, writeFileSync } from "node:fs";
import { resolve } from "node:path";
import { pathToFileURL } from "node:url";

const pct = (xs, p) => xs[Math.min(xs.length - 1, Math.floor(xs.length * p))];

async function run(modulePath, baseURL, textsPath, out) {
  const t0 = performance.now();
  const sdk = await import(pathToFileURL(resolve(modulePath)).href);
  const importMs = performance.now() - t0;
  const client = new sdk.TypeSafeClient({ apiKey: "none", baseURL });
  const questions = {
    billing: sdk.noul("Is this about billing or payments?"),
    intent: sdk.choice("What does the writer want?", {
      help: "they need help or support",
      complain: "they are unhappy",
      inform: "they are sharing information",
    }),
    urgency: sdk.score("How urgent is this?", ["not urgent", "somewhat urgent", "very urgent"]),
  };
  const texts = JSON.parse(readFileSync(textsPath, "utf8"));
  const models = (await client.models.list()).map((m) => m.name);
  const rows = [];
  const took = [];
  for (const [i, text] of texts.entries()) {
    const start = performance.now();
    const r = await client.systemOne({ state: text, questions, model: "jev-latest" });
    if (i) took.push(performance.now() - start);
    rows.push(r.answers);
  }
  const listed = [];
  for (let i = 0; i < 2000; i++) {
    const start = performance.now();
    await client.models.list();
    listed.push(performance.now() - start);
  }
  took.sort((a, b) => a - b);
  listed.sort((a, b) => a - b);
  const summary = {
    client: modulePath,
    runtime: typeof Bun !== "undefined" ? `bun ${Bun.version}` : typeof Deno !== "undefined" ? `deno ${Deno.version.deno}` : `node ${process.version}`,
    models,
    calls: texts.length,
    import_ms: +importMs.toFixed(2),
    p50_ms: +pct(took, 0.5).toFixed(2),
    p99_ms: +pct(took, 0.99).toFixed(2),
    models_p50_us: +(pct(listed, 0.5) * 1e3).toFixed(1),
    models_p99_us: +(pct(listed, 0.99) * 1e3).toFixed(1),
    rss_mb: typeof process !== "undefined" ? +(process.memoryUsage().rss / 2 ** 20).toFixed(1) : null,
  };
  console.log(JSON.stringify(summary));
  writeFileSync(out, JSON.stringify({ summary, answers: rows }));
}

function compare(paths) {
  const [base, ...others] = paths.map((p) => JSON.parse(readFileSync(p, "utf8")));
  for (const other of others) {
    const same = base.answers.filter((a, i) => JSON.stringify(a) === JSON.stringify(other.answers[i])).length;
    console.log(`${base.summary.client} vs ${other.summary.client}: ${same} of ${base.answers.length} responses identical`);
  }
}

const [cmd, ...rest] = process.argv.slice(2);
if (cmd === "compare") compare(rest);
else await run(...rest);

File: tools/python/typesafe.py
"""Runs the same System One calls through `typesafe_sdk` or `kime.typesafe` and writes the
answers and the time each call took, so the two clients can be compared answer for answer.

    uv run --no-project --with typesafe-sdk==0.7.1 python tools/python/typesafe.py sdk http://127.0.0.1:18831 texts.json out-sdk.json
    python tools/python/typesafe.py kime http://127.0.0.1:18831 texts.json out-kime.json
    python tools/python/typesafe.py local /path/to/laya texts.json out-local.json
    python tools/python/typesafe.py compare out-sdk.json out-kime.json out-local.json

`texts.json` is a list of strings. Each text is asked a noul, a choice and a score question, and
the first call of each run is left out of the timings.
"""

import json
import statistics
import sys
import time


def questions(t):
    return {
        "billing": t.Noul(instructions="Is this about billing or payments?"),
        "intent": t.Choice(
            instructions="What does the writer want?",
            criteria={"help": "they need help or support", "complain": "they are unhappy", "inform": "they are sharing information"},
        ),
        "urgency": t.Score(instructions="How urgent is this?", criteria=["not urgent", "somewhat urgent", "very urgent"]),
    }


def run(kind, target, texts, out):
    if kind == "sdk":
        import typesafe_sdk as t

        client = t.TypeSafeClient(api_key="none", base_url=target)
    else:
        import kime.typesafe as t

        if kind == "local":
            client = t.TypeSafeClient(local=target, device="cpu", precision="f32")
        else:
            client = t.TypeSafeClient(base_url=target)
    q = questions(t)
    models = [m.name for m in client.models.list().models]
    rows, took = [], []
    for i, text in enumerate(texts):
        start = time.perf_counter()
        r = client.system_one(text, q, model="jev-latest")
        if i:
            took.append(time.perf_counter() - start)
        rows.append({k: a.model_dump() for k, a in r.answers.items()})
    ms = sorted(x * 1e3 for x in took)
    summary = {"client": kind, "models": models, "calls": len(texts), "mean_ms": statistics.fmean(ms),
               "p50_ms": ms[len(ms) // 2], "p99_ms": ms[int(len(ms) * 0.99)]}
    print(json.dumps(summary))
    with open(out, "w") as f:
        json.dump({"summary": summary, "answers": rows}, f)


def compare(paths):
    runs = [json.load(open(p)) for p in paths]
    base = runs[0]
    for other in runs[1:]:
        same = sum(a == b for a, b in zip(base["answers"], other["answers"]))
        print("%s vs %s: %d of %d responses identical" % (
            base["summary"]["client"], other["summary"]["client"], same, len(base["answers"])))


if __name__ == "__main__":
    if sys.argv[1] == "compare":
        compare(sys.argv[2:])
    else:
        run(sys.argv[1], sys.argv[2], json.load(open(sys.argv[3])), sys.argv[4])
